### PR TITLE
Change min zoom for landuse, POIs to be closer to Bubble Wrap

### PR DESCRIPTION
**Pull request by zerebubuth**
Changes min zoom:

1. For `landuse` to match the Bubble Wrap config and adds `tier` parameter.
2. For `pois` to restrict existing `min_zoom` to be no lower than the corresponding `landuse` value, and add `tier` parameter.

Connects to #473.

@rmarianski could you review, please?

@nvkelso the changes restricted many of the zooms for POIs, especially those from points - please could you check the changes to the integration tests match the desired behaviour?

### CODE REVIEW

**Comment by zerebubuth** [comment]
I forgot to mention: Also added the ability for `min_zoom` in the YAML config to be an expression. This is useful for combining two sets of requirements for min zoom.

**Comment by nvkelso** [comment]
I'll have a look Monday.

Question:

We had:

-  - filter: {natural: forest}
-    min_zoom: GREATEST( LEAST(zoom + 2, 14), 12)

But some of the changes and comments make it seem like nodes only show up at 16 now instead of the earlier custom min_zoom? Can you walk me through an example, please?

**Comment by zerebubuth** [comment]
> But some of the changes and comments make it seem like nodes only show up at 16 now instead of the earlier custom min_zoom? Can you walk me through an example, please?

Sure. Take [Santa Rosa Ranget District](http://www.openstreetmap.org/node/357559979) as an example - it's [a forest landuse POI from the tests](https://github.com/tilezen/vector-datasource/pull/997/files#diff-b74f5b6546f151b0946770fe96922637R58), and would have previously appeared at z14.

According to the [rules from Bubble Wrap](https://github.com/tilezen/vector-datasource/issues/473#issuecomment-170080923), because it has an area of zero (it's a point), we only start to show it at zoom 16.

**Comment by nvkelso** [comment]
If a feature doesn’t have an area, we should prefer the existing node min_zoom. Here’s Olga’s PR: https://github.com/tilezen/vector-datasource/pull/939/files. (They shouldn’t all default to 16, they should still have their existing min_zoom from landuse and pois queries.)

**Comment by nvkelso** [comment]
Having a look...

**Comment by zerebubuth** [comment]
Thanks!

File: integration-test/473-landuse-tier.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# http://www.openstreetmap.org/way/167274589
+# area 300363008
+assert_has_feature(
+    16, 10818, 21900, 'landuse',
```
Comment (nvkelso) [inline integration-test/473-landuse-tier.py]: Should this test be for zoom 4 (the first zoom which would include this feature)?

File: integration-test/473-landuse-tier.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      'min_zoom': 3 })
+assert_has_feature(
+    6, 10, 21, 'pois',
+    { 'kind': 'national_park', 'id': 167274589, 'tier': 1,
+      'min_zoom': 6.67 })
+
+# http://www.openstreetmap.org/relation/921675
+# area 30089300
+assert_has_feature(
+    16, 14579, 29651, 'landuse',
```
Comment (nvkelso) [inline integration-test/473-landuse-tier.py]: Should this test be for zoom 8?

File: integration-test/473-landuse-tier.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+assert_has_feature(
+    8, 56, 115, 'pois',
+    { 'kind': 'national_park', 'id': -921675, 'tier': 1,
+      'min_zoom': 8.33 })
+
+# this is USFS, so demoted to tier 2 :-(
+# http://www.openstreetmap.org/way/34416231
+# area 86685400
+assert_has_feature(
+    16, 18270, 25157, 'landuse',
```
Comment (nvkelso) [inline integration-test/473-landuse-tier.py]: Should this test be for zoom 8?

File: integration-test/742-predictable-layers-pois.py [hunk 4]
```diff
@@ -87,12 +87,12 @@
...
 # http://www.openstreetmap.org/way/30903221
 assert_has_feature(
-    12, 654, 1583, 'pois',
+    13, 1308, 3166, 'pois',
     {'id': 30903221, 'kind': 'golf_course'})
 
 # Label placement Golf_course in landuse
 assert_no_matching_feature(
-    12, 654, 1583, 'landuse',
+    13, 1308, 3166, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 5]
```diff
@@ -104,12 +104,12 @@
...
 # http://www.openstreetmap.org/way/330274727
 assert_has_feature(
-    11, 327, 793, 'pois',
+    13, 1308, 3174, 'pois',
     {'id': 330274727, 'kind': 'military'})
 
 # Label placement military in landuse
 assert_no_matching_feature(
-    11, 327, 793, 'landuse',
+    13, 1308, 3174, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 6]
```diff
@@ -155,12 +155,12 @@
...
 # http://www.openstreetmap.org/way/23871270
 assert_has_feature(
-    11, 327, 791, 'pois',
+    13, 1310, 3166, 'pois',
     {'id': 23871270, 'kind': 'park'})
 
 # Label placement park in landuse
 assert_no_matching_feature(
-    11, 327, 791, 'landuse',
+    13, 1310, 3166, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 7]
```diff
@@ -172,12 +172,12 @@
...
 # http://www.openstreetmap.org/way/26278098
 assert_has_feature(
-    13, 1308, 3167, 'pois',
+    14, 2617, 6334, 'pois',
     {'id': 26278098, 'kind': 'plant'})
 
 # Label placement plant in landuse
 assert_no_matching_feature(
-    13, 1308, 3167, 'landuse',
+    14, 2617, 6334, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 8]
```diff
@@ -269,12 +269,12 @@
...
 assert_has_feature(
-    12, 655, 1583, 'pois',
-    {'id': 239634932, 'kind': 'wastewater_plant', 'min_zoom': 12})
+    13, 1310, 3167, 'pois',
+    {'id': 239634932, 'kind': 'wastewater_plant', 'min_zoom': 13})
 
 # Label placement wastewater_plant in landuse
 assert_no_matching_feature(
-    12, 655, 1583, 'landuse',
+    13, 1310, 3167, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 9]
```diff
@@ -320,23 +320,23 @@
...
 # http://www.openstreetmap.org/way/207859675
 assert_has_feature(
-    13, 1413, 3274, 'pois',
+    14, 2826, 6549, 'pois',
     {'id': 207859675, 'kind': 'wood'})
 
 # Label placement landuse: wood in landuse
 assert_no_matching_feature(
-    13, 1413, 3274, 'landuse',
+    14, 2826, 6549, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 10]
```diff
@@ -320,23 +320,23 @@
...
 # http://www.openstreetmap.org/way/372445925
 assert_has_feature(
-    13, 1309, 3165, 'pois',
+    14, 2618, 6330, 'pois',
     {'id': 372445925, 'kind': 'natural_wood'})
 
 # Label placement natural: wood in landuse
 assert_no_matching_feature(
-    13, 1309, 3165, 'landuse',
+    14, 2618, 6330, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: integration-test/742-predictable-layers-pois.py [hunk 11]
```diff
@@ -348,12 +348,12 @@
...
 assert_has_feature(
-    12, 714, 1623, 'pois',
-    {'id': 164878781, 'kind': 'works', 'min_zoom': 12})
+    13, 1429, 3247, 'pois',
+    {'id': 164878781, 'kind': 'works', 'min_zoom': 13})
 
 # Label placement works in landuse
 assert_no_matching_feature(
-    12, 714, 1623, 'landuse',
+    13, 1429, 3247, 'landuse',
```
Comment (nvkelso) [inline integration-test/742-predictable-layers-pois.py]: My changes in #1015 only generate landuse label placements starting at zoom 15.

File: yaml/landuse.yaml [hunk 12]
```diff
@@ -1,114 +1,482 @@
...
+  # wilderness_hut
+  - filter: {tourism: wilderness_hut}
+    min_zoom: *tier6_min_zoom
+    output: {kind: wilderness_hut, tier: 6}
+  # hanami
+  - filter: {tourism: hanami}
+    min_zoom: *tier6_min_zoom
+    output: {kind: hanami, tier: 6}
+  # tree_row
+  - filter: {natural: tree_row}
```
Comment (nvkelso) [inline yaml/landuse.yaml]: `tree_row` should not be part of tier 6 (or any tier).

File: yaml/landuse.yaml [hunk 13]
```diff
@@ -1,114 +1,482 @@
...
+    min_zoom: *tier6_min_zoom
+    output: {kind: tree_row, tier: 6}
+  # hedge
+  - filter: {barrier: hedge}
```
Comment (nvkelso) [inline yaml/landuse.yaml]: `hedge` should not be part of tier 6 (or any tier).

File: yaml/landuse.yaml [hunk 14]
```diff
@@ -1,114 +1,482 @@
...
+           WHEN way_area >=  5000 THEN 15
+           ELSE                        16
+      END
+    output: {kind: parking, tier: 6}
+
+
+  ############################################################
+  # NOT IN ANY TIER
+  ############################################################
+  # park w/ US Parks Service operator
```
Comment (nvkelso) [inline yaml/landuse.yaml]: `park` or `national_park` with US National Park Service operator?

File: yaml/pois.yaml [hunk 15]
```diff
@@ -1,16 +1,424 @@
...
+    output:
+      kind: national_park
+      tier: 1
+
+  ############################################################
+  # TIER 2
+  ############################################################
+  - filter:
+      boundary: national_park
+      not: {operator: *us_parks_service}
```
Comment (nvkelso) [inline yaml/pois.yaml]: Agree this basic `national_park` logic is fine for this PR, to be refined in https://github.com/tilezen/vector-datasource/issues/987.

File: yaml/landuse.yaml [hunk 16]
```diff
@@ -1,114 +1,482 @@
...
+    output: { kind: natural_forest, tier: 2 }
+
+  - filter: { natural: wood, operator: *us_forest_service }
+    min_zoom: { max: [ 6, { lit: *tier2_min_zoom } ] }
+    output: { kind: natural_wood, tier: 2 }
+  - filter: { natural: wood }
+    min_zoom: { max: [ 10, { lit: *tier2_min_zoom } ] }
+    output: { kind: natural_wood, tier: 2 }
+
+  # TODO: conservation? doesn't seem to be an existing kind.
```
Comment (nvkelso) [inline yaml/landuse.yaml]: Agree `conservation` is bogus, let's remove reference to it.

File: yaml/pois.yaml [hunk 17]
```diff
@@ -1,16 +1,424 @@
...
+  - filter: {natural: forest}
+    min_zoom: { min: [ { max: [ 12, { lit: zoom + 2 }, { lit: *tier2_min_zoom } ] }, 14 ] }
+    output: { kind: natural_forest, tier: 2 }
+
+  # natural woods - note already clampted to 12
+  - filter: {natural: wood}
+    min_zoom: { min: [ { max: [ 12, { lit: zoom + 2 }, { lit: *tier2_min_zoom } ] }, 14 ] }
+    output: { kind: natural_wood, tier: 2 }
+
+  # TODO: conservation? doesn't seem to be an existing kind.
```
Comment (nvkelso) [inline yaml/pois.yaml]: Agree `conservation` is bogus, let's remove reference to it.

File: yaml/landuse.yaml [hunk 18]
```diff
@@ -1,114 +1,482 @@
...
+    output:
+      kind: national_park
+      tier: 1
+
+  ############################################################
+  # TIER 2
+  ############################################################
+  - filter:
+      boundary: national_park
+      not: { operator: *us_parks_service }
```
Comment (nvkelso) [inline yaml/landuse.yaml]: My filters include NOT for both `*us_parks_service` and `*us_forest_service`, but I think https://github.com/tilezen/vector-datasource/issues/987 is a better place to track that refinement.

File: yaml/landuse.yaml [hunk 19]
```diff
@@ -1,114 +1,482 @@
...
+    min_zoom: *tier2_min_zoom
+    output:
+      kind: national_park
+      tier: 2
+  - filter:
+      any:
+        leisure: park
+        landuse: park
+      not: { operator: *us_parks_service }
```
Comment (nvkelso) [inline yaml/landuse.yaml]: My filters include NOT for both `*us_parks_service` and `*us_forest_service`, but I think https://github.com/tilezen/vector-datasource/issues/987 is a better place to track that refinement.

File: yaml/landuse.yaml [hunk 20]
```diff
@@ -1,114 +1,482 @@
...
+    min_zoom: *tier2_min_zoom
+    output:
+      kind: park
+      tier: 2
+
+  - filter: { landuse: forest, operator: *us_forest_service }
```
Comment (nvkelso) [inline yaml/landuse.yaml]: My filters no longer include `*us_forest_service` for `forest` features, but I think https://github.com/tilezen/vector-datasource/issues/987 is a better place to track that refinement.

File: yaml/landuse.yaml [hunk 21]
```diff
@@ -1,114 +1,482 @@
...
+  - filter: {attraction: maze}
+    min_zoom: *tier5_min_zoom
+    output: {kind: maze, tier: 5}
+  # beach
+  - filter: {natural: beach}
+    min_zoom: *tier5_min_zoom
+    output: {kind: beach, tier: 5}
+  # glacier
+  - filter: {natural: glacier}
+    min_zoom: *tier5_min_zoom
```
Comment (nvkelso) [inline yaml/landuse.yaml]: `glacier` wasn't in a tier in either Walkabout or Bubble Wrap, although labels are currently only showing up at zoom 10. I propose `tier3_min_zoom` for `glacier` landuse polygons instead.

File: yaml/landuse.yaml [hunk 22]
```diff
@@ -1,114 +1,482 @@
...
+    output: { kind: natural_forest, tier: 2 }
+  - filter: { natural: forest }
+    min_zoom: { max: [ 10, { lit: *tier2_min_zoom } ] }
+    output: { kind: natural_forest, tier: 2 }
+
+  - filter: { natural: wood, operator: *us_forest_service }
+    min_zoom: { max: [ 6, { lit: *tier2_min_zoom } ] }
+    output: { kind: natural_wood, tier: 2 }
+  - filter: { natural: wood }
+    min_zoom: { max: [ 10, { lit: *tier2_min_zoom } ] }
```
Comment (nvkelso) [inline yaml/landuse.yaml]: This was `14` (instead of `10`).

File: yaml/landuse.yaml [hunk 23]
```diff
@@ -1,114 +1,482 @@
...
+    output: { kind: farmland, tier: 2 }
+
+  ############################################################
+  # TIER 3
+  ############################################################
+  - filter: {aeroway: aerodrome}
+    min_zoom: *tier3_min_zoom
+    output: {kind: aerodrome, tier: 3}
+  - filter: {landuse: military}
+    min_zoom: *tier3_min_zoom
```
Comment (nvkelso) [inline yaml/landuse.yaml]: Agree with simplified logic for `military`.

File: yaml/pois.yaml [hunk 24]
```diff
@@ -1,16 +1,424 @@
+#
+# NOTE: these globals are shared between the POIs and landuse. if you make an
```
Comment (nvkelso) [inline yaml/pois.yaml]: **A+**

File: yaml/pois.yaml [hunk 25]
```diff
@@ -1,16 +1,424 @@
...
+  # forests
+  - filter:
+      landuse: forest
+      protect_class: '6'
+    min_zoom: { min: [ { max: [ 7, { lit: zoom + 2 }, { lit: *tier2_min_zoom } ] }, 14 ] }
+    output: { kind: forest, tier: 2 }
+  - filter:
+      landuse: forest
+      operator: *us_forest_service
+    min_zoom: { min: [ { max: [ 9, { lit: zoom + 2 }, { lit: *tier2_min_zoom } ] }, 14 ] }
```
Comment (nvkelso) [inline yaml/pois.yaml]: If the operator is `*us_forest_service` then I think we'd want to see them at zoom 7 like the rule above?

File: yaml/pois.yaml [hunk 26]
```diff
@@ -1,16 +1,424 @@
...
+  # farm
+  - filter: {landuse: farm}
+    min_zoom: { max: [ 15, { lit: *tier2_min_zoom } ] }
+    output: { kind: farm, tier: 2 }
+  # urban, rural, residential, farmland - no POIs
+
+  ############################################################
+  # TIER 3
+  ############################################################
+  # aerodrome - no POIs
```
Comment (nvkelso) [inline yaml/pois.yaml]: Was expecting to see something like:

```
  - filter:
      aeroway: [aerodrome, airport]
    min_zoom: LEAST(zoom + 4.12, 13)
    output:
      kind: {col: aeroway}
```

File: yaml/pois.yaml [hunk 27]
```diff
@@ -1,16 +1,424 @@
...
+  - filter: {leisure: stadium}
+    min_zoom: { min: [ { max: [ { lit: zoom + 2.3 }, { lit: *tier4_min_zoom } ] }, 15 ] }
+    output: {kind: stadium, tier: 4}
+  # substation
+  - filter: {power: substation}
+    min_zoom: 15
+    output:
+      kind: substation
+      tier: 4
+  # tower - no POI
```
Comment (nvkelso) [inline yaml/pois.yaml]: Was expecting to see something for tower outside of the tiers:

```
  - filter: {power: tower}
    min_zoom: 16
    output: {kind: power_tower}
```

File: yaml/pois.yaml [hunk 28]
```diff
@@ -1,16 +1,424 @@
...
+    output:
+      kind: works
+      tier: 4
+  # water_works
+  - filter: {man_made: water_works}
+    min_zoom: { min: [ { max: [ 12, { lit: zoom + 3 }, { lit: *tier4_min_zoom } ] }, 14 ] }
+    output:
+      kind: water_works
+      tier: 4
+  # wildlife_park - no POI
```
Comment (nvkelso) [inline yaml/pois.yaml]: Was expecting to see something like (not in a tier):

```
  - filter:
      tags->zoo: [enclosure, petting_zoo, aviary, wildlife_park]
    min_zoom: 17
    output:
      kind: {col: tags->zoo}
```

File: yaml/pois.yaml [hunk 29]
```diff
@@ -1,16 +1,424 @@
...
+  ############################################################
+  # parking
+  - filter: {amenity: parking}
+    min_zoom: 17
+    output: {kind: parking, tier: 6}
+
+  ############################################################
+  # NOT IN ANY TIER
+  ############################################################
+  - filter: {historic: battlefield, operator: *us_forest_service}
```
Comment (nvkelso) [inline yaml/pois.yaml]: ¿us_forest_service doesn't operate any battlefields (but national park service does)?

Comment (nvkelso) [inline yaml/pois.yaml]: (Probably just remove the operator part of the filter.)

File: yaml/pois.yaml [hunk 30]
```diff
@@ -1,16 +1,424 @@
...
+    min_zoom: GREATEST(10, LEAST(zoom + 4.0, 17))
+    output: {kind: battlefield}
+  # park w/ US Parks Service operator
+  - filter:
+      any:
+        leisure: park
+        landuse: park
+    min_zoom: GREATEST( LEAST(zoom + 2, 14), 9)
```
Comment (nvkelso) [inline yaml/pois.yaml]: Should this be in `tier2_min_zoom`?

File: yaml/pois.yaml [hunk 31]
```diff
@@ -1,16 +1,424 @@
...
+    output: { kind: park }
+
   - filter:
       aeroway: [aerodrome, airport]
```
Comment (nvkelso) [inline yaml/pois.yaml]: Here's the other aerodrome, airport reference!

File: yaml/pois.yaml [hunk 32]
```diff
@@ -674,102 +1007,28 @@ filters:
...
-    min_zoom: GREATEST( LEAST(zoom + 2, 14), 12)
-    output:
-      kind: natural_wood
-  - filter: {man_made: works}
-    min_zoom: GREATEST( LEAST(zoom + 3, 14), 12)
+
+  ############################################################
+  # TIER 6 EXTRA - ATTRACTION
+  #
+  # Because this is so generic, we want other tags to match
```
Comment (nvkelso) [inline yaml/pois.yaml]: **A+**

**Comment by nvkelso** [comment]
Overall looks really good :) Thank you for putting so much time into this one!

The tiers are useful to organize the filters, but they feel like an internal `mz_*` variable that perhaps shouldn't be exported in tiles? Don't take them out just yet, though as I think they'll be useful to QA the changes on dev to really see.

File: yaml/landuse.yaml (hunk 22 above)
Comment (zerebubuth) [inline yaml/landuse.yaml]: Just checking: you want natural woods without USFS operator to show up only at zoom 14?

File: yaml/pois.yaml (hunk 25 above)
Comment (zerebubuth) [inline yaml/pois.yaml]: That wasn't how the original logic worked. Easy change, though.

File: yaml/pois.yaml (hunk 27 above)
Comment (zerebubuth) [inline yaml/pois.yaml]: The corresponding feature in `landuse` isn't a `power=tower`, but a `man_made=tower`. Do we want to conflate the two?

File: yaml/pois.yaml (hunk 30 above)
Comment (zerebubuth) [inline yaml/pois.yaml]: The original logic excluded parks with the US Parks Service operator from the tier, so this is the "else" clause to ensure that things not caught by previous expressions still get assigned a `kind`, if not a `tier`.

**Comment by zerebubuth** [comment]
I pushed a bunch of fixes for several of the comments. @nvkelso could you give it another look-over, please?

File: yaml/pois.yaml (hunk 17 above)
Comment (nvkelso) [inline yaml/pois.yaml]: This TODO should be removed.

File: yaml/landuse.yaml (hunk 22 above)
Comment (nvkelso) [inline yaml/landuse.yaml]: Thanks for checking. I think the 14 before was an error that you corrected to 10 (to be consistent with forest). Good catch!

File: yaml/pois.yaml (hunk 30 above)
Comment (nvkelso) [inline yaml/pois.yaml]: Thanks for the explanation, sounds good!

**Comment by nvkelso** [comment]
Remaining tasks:

- another landuse labels to zoom 15: https://github.com/tilezen/vector-datasource/pull/997/files#r77759577
- `wildlife_park` nodes: https://github.com/tilezen/vector-datasource/pull/997/files#r77763317

All other changes verified, thanks!

**Comment by nvkelso** [comment]
(I think the bad reference to conservation was in there because there's an OSM page for that Conservation links to all the more specific tags.)

**Comment by zerebubuth** [comment]
> another landuse labels to zoom 15: https://github.com/tilezen/vector-datasource/pull/997/files#r77759577

Well spotted! I missed that, but fixed in ac7f517. Had a look through for others, too... I'm pretty confident we got them all this time.

> `wildlife_park` nodes: https://github.com/tilezen/vector-datasource/pull/997/files#r77763317

They should be in there, just [under the comment](https://github.com/tilezen/vector-datasource/pull/997/files#diff-85e6f75825fd3fd79970e1da0dd6cdaeR308). Was that what you meant, or should that filter be present somewhere else?

File: yaml/pois.yaml (hunk 28 above)
Comment (nvkelso) [inline yaml/pois.yaml]: This is actually in there, right below this comment!

**Comment by nvkelso** [comment]
Oops, my bad about `wildlife_park`.

👍 ready to merge.

**Pull request merged**